scalyr_agent/line_matcher.py

```diff
@@ -125,6 +169,135 @@ def _readline( self, file_like, max_length=0 ):
+        if line_len > max_length:
+            line = line[:max_length]
+            line += "...truncated %d bytes..." % (line_len - max_length)
+        return LogLine( line )
+
+    def _readline( self, file_like, max_length=0, current_file=None ):
+        """ Takes a file_like object (e.g. anything conforming to python's file interface
+        and attempts to parse the line as JSON.
+
+        This method will attempt to process the line as json, extracting a log message, a timestamp and any
+        other remaining fields, all of which are returned as a LogLine object.
+
+        When parsing as json, each line is required to be a fully formed json object, otherwise the full contents of
+        the line will be returned unprocessed.
+
+        The configuration options 'json_message_field' and 'json_timestamp_field' are used to specify which field to use
+        as the log message, and which field to use for the timestamp.
+
+        'json_message_field' defaults to 'log' and if no field with this name is found, the function will return the
+        full line unprocessed.
+
+        'json_timestamp_field' defaults to 'time', and the value of this field is required to be in rfc3339 format
+        otherwise an error will occur.  If the field does not exist in the json object, then the agent will use
+        the current time instead.  Note, the value specified in the timestamp field might not be the final value uploaded
+        to the server, as the agent ensures that the timestamps of all messages are monotonically increasing.  For
+        this reason, if the timestamp field is found, a 'raw_timestamp' attributed is also added to the LogLine's attrs.
+
+        All other fields of the json object will be stored in the attrs dict of the LogLine object.
+
+        The method will read up to `read_page_size` bytes from the buffer, and truncate any log message to `max_line_size` bytes.
+
+        @param file_like: a file like object (e.g. StringIO)
+        @param max_length: the maximum length to read from the file_like object
+
+        @return: a tuple containing the content read from the file_like object and whether
+            this is a partial line or not.  `partial` will only be true if `read_page_size`
+            was exceeded and a full json object wasn't found.
+
+        @rtype: (LogLine, bool)
+        """
+
+        # set some sensible defaults
+        if max_length == 0:
+            max_length = self.read_page_size
+
+        if current_file is None:
+            current_file = '<unknown>'
+
+        # read from the buffer
+        line = file_like.readline( max_length )
+
+        #return early if no line found
+        if len( line ) == 0:
+            return LogLine( '' ), False
+
+        # check for presence of newline chars to see if this is a partial line
+        partial = line[-1] != '\n' and line[-1] != '\r'
+
+        # early abort if this appears to be a partial line
+        if partial:
```

Comment:
What's going to happen after we truncate this line?  We still haven't seen the end of the JSON object, so I think we are going to keep reading the buffer and discarding it, in what, max line chunks?  Or is it in the page size chunk?

I do wonder if we should do something better here like keep reading the buffer (in future calls to this method) where we just keep looking for the end of a JSON object .. throwing away everything as we do.  Essentially, I'd like to emit just one bad line and then pick up at the next valid line.